### PR TITLE
[CSSolver] Add closure parameter type caching

### PR DESCRIPTION
**Pull request by xedin**
While inferring avoid associating type variables with closure
parameters, use cache instead and only set types when everything
is properly type-checked, this avoids multiple problems one of
them - leaking type variables outside of constraint system they
belong to.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeChecker.h [hunk 1]
```diff
@@ -1929,11 +1929,14 @@ class TypeChecker final : public LazyResolver {
...
   ///
+  /// \param valueType The optional type associated with given value, if it's
+  /// not set, type of the declaration is going to be used instead.
+  ///
   /// \param base The optional base expression of this value reference
   ///
   /// \param wantInterfaceType Whether we want the interface type, if available.
   Type getUnopenedTypeOfReference(VarDecl *value, Type baseType,
-                                  DeclContext *UseDC,
+                                  DeclContext *UseDC, Type valueType = Type(),
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/TypeChecker.h]: Yeah, that's what I was thinking as well, just wanted to check preliminary how much a broke :)

**Comment by slavapestov** [comment]
Once you get it working, you can change ValueDecl::setInterfaceType() to assert that the new type has no type variables in it, and finally remove the ExprCleanser and CleanupIllFormedExpressionRAII.

File: lib/Sema/ConstraintSystem.h [hunk 2]
```diff
@@ -1622,6 +1634,19 @@ class ConstraintSystem {
     return TypeLocTypes.find(&L)->second;
   }
 
+  Type getType(const ValueDecl *P, bool queryDeclOnMiss = true) const {
```
Comment (rudkx) [inline lib/Sema/ConstraintSystem.h]: Can I suggest splitting this into three functions?
  `getType(const ParamDecl *)` - asserts hasType() and returns the type

  `getTypeOrNull(const ValueDecl *D)`:
```
      if (auto *P = dyn_cast<ParamDecl>(D)) 
        return getType(P);

      return Type();
```

`getTypeOrInterfaceType`: Your current function, minus the argument for `queryDeclOnMiss`

Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Sure!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
@shahmishal `build-script` keeps crashing aster running `TestAffineTransform.test_NSCoding` on Linux, is this a known problem?

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by rudkx** [comment]
@xedin At first glance this looks good. Can you ping me when the merge conflict is resolved and tests are passing, and I'll take a closer look?

**Comment by xedin** [comment]
@swift-ci please clean test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@shahmishal Are there any logs to help figure out what is build-script crashing in `TestFoundation`?

File: lib/Sema/ConstraintSystem.cpp [hunk 3]
```diff
@@ -770,20 +770,22 @@ static bool doesStorageProduceLValue(TypeChecker &TC,
...
+                                                  bool wantInterfaceType) {
+  return TC.getUnopenedTypeOfReference(
+      value, baseType, UseDC,
+      [&](VarDecl *var) -> Type { return getType(var, wantInterfaceType); },
+      base, wantInterfaceType);
+}
 
-  requestedType = requestedType->getWithoutSpecifierType()
-    ->getReferenceStorageReferent();
+Type TypeChecker::getUnopenedTypeOfReference(
```
Comment (slavapestov) [inline lib/Sema/ConstraintSystem.cpp]: Does this need to exist? Why not always use the ConstraintSystem version

Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: `TypeChecker::buildCheckedRefExpr` is still using version from TypeChecker and we don't have constraint system it those cases :(

**Comment by xedin** [comment]
Linux failure is definitely something related to my changes, but I don't yet understand how because for test-case in question type-checker produces the same solution and type-checked AST...

**Comment by xedin** [comment]
@swift-ci please ASAN test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
This time it's seems like an infra problem :)

```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/Output/swift-build-self-host.py.tmp.dir/swiftpm/Sources/SPMLLBuild/llbuild.swift:15:19: error: no such module 'llbuild'
18:32:44 @_exported import llbuild
18:32:44                   ^
```

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Commits got merged but for some reason `Github` failed to mark this PR as merged...

**Pull request closed without merging**